File: solutions/c/space-age/2/space_age.c
#include "space_age.h"
#include "space_age.h"

#define EARTH_YEAR_IN_SECONDS 31557600.0f

static const float planet_years_per_second[] = {
    1.0f / (0.2408467f   * EARTH_YEAR_IN_SECONDS), // Mercury
    1.0f / (0.61519726f  * EARTH_YEAR_IN_SECONDS), // Venus
    1.0f / (1.0f         * EARTH_YEAR_IN_SECONDS), // Earth
    1.0f / (1.8808158f   * EARTH_YEAR_IN_SECONDS), // Mars
    1.0f / (11.862615f   * EARTH_YEAR_IN_SECONDS), // Jupiter
    1.0f / (29.447498f   * EARTH_YEAR_IN_SECONDS), // Saturn
    1.0f / (84.016846f   * EARTH_YEAR_IN_SECONDS), // Uranus
    1.0f / (164.79132f   * EARTH_YEAR_IN_SECONDS), // Neptune
};

float age(planet_t planet, int64_t seconds) {
    if (planet < MERCURY || planet > NEPTUNE) return -1.0f;
    return seconds * planet_years_per_second[planet];
}

File: solutions/elixir/space-age/2/lib/space_age.ex
defmodule SpaceAge do

  @type planet :: :mercury | :venus | :earth | :mars
                | :jupiter | :saturn | :uranus | :neptune

  @earth_year_seconds 365.25 * 24 * 60 * 60

  @orbital_periods %{
    mercury: 0.2408467,
    venus:   0.61519726,
    earth:   1.0,
    mars:    1.8808158,
    jupiter: 11.862615,
    saturn:  29.447498,
    uranus:  84.016846,
    neptune: 164.79132
  }

  for {planet, period} <- @orbital_periods do
    @spec age_on(unquote(planet), non_neg_integer()) :: {:ok, float()}
    def age_on(unquote(planet), seconds)
        when is_integer(seconds) and seconds >= 0 do
      {:ok, seconds / @earth_year_seconds / unquote(period)}
    end
  end

  @spec age_on(atom(), non_neg_integer()) :: {:error, String.t()}
  def age_on(_planet, seconds)
      when is_integer(seconds) and seconds >= 0 do
    {:error, "not a planet"}
  end
  
end
